### PR TITLE
fix: disable timestamps when CTC weights are missing

### LINKED ISSUE #3208: [Fun-ASR-MLT-Nano-2512] Official checkpoint omits CTC weights but still returns invalid timestamps
## Summary

The official ModelScope checkpoint for `FunAudioLLM/Fun-ASR-MLT-Nano-2512`
contains no trained CTC decoder/head weights. FunASR initializes those layers
randomly, logs missing-key warnings, and still returns structurally valid but
semantically invalid `timestamps`.

Text transcription itself is accurate for English, Japanese, Korean, and Thai.
The issue is specifically the timestamp contract.

## Environment

- FunASR: 1.3.14
- Model: `FunAudioLLM/Fun-ASR-MLT-Nano-2512`
- Source: official ModelScope `master` checkpoint
- `model.pt` size: 1,971,162,898 bytes
- `model.pt` SHA-256:
  `f7ddbef5ff8c57254e6a045f0b316501b704fe2b774f962efc6e402364e8fe37`
- Device used for reproduction: CPU
- Input: clean TTS speech, approximately 7-11 seconds per sample

## Checkpoint inspection

```python
import torch

checkpoint = torch.load(
    "model.pt",
    map_location="meta",
    weights_only=False,
    mmap=True,
)
state = checkpoint["state_dict"]

print(len(state))
print(sum(key.startswith("ctc_decoder.") for key in state))
print(sum(key.startswith("ctc.") for key in state))
```

Output:

```text
1261
0
0
```

During `AutoModel(...)` initialization, FunASR reports 84 missing
`ctc_decoder.*` keys and 2 missing `ctc.*` keys.

## Minimal inference path

```python
from funasr import AutoModel

model = AutoModel(
    model="FunAudioLLM/Fun-ASR-MLT-Nano-2512",
    device="cpu",
    disable_update=True,
)

result = model.generate(
    input=["clean_tts.wav"],
    cache={},
    batch_size=1,
    language="英文",
    itn=True,
)

print(result[0]["text"])
print(result[0]["timestamps"][:3])
```

## Observed results

Transcription similarity against the exact TTS source text:

| Language | Similarity |
|---|---:|
| Japanese | 1.0000 |
| Korean | 0.9792 |
| English | 1.0000 |
| Thai | 0.9947 |

The returned timestamps are not usable:

- Every returned timestamp has `score=0.0`.
- Japanese: one character receives a 3.36-second span.
- Korean: one character receives a 7.14-second span.
- English: the token `blush` receives a 6.06-second span in a 7.46-second clip.
- Thai: one token receives a 6.18-second span.
- Timestamps remain monotonic and within the audio duration, so a structural
  validation alone does not detect the failure.

This is consistent with forced alignment running against a randomly initialized
CTC head.

## Checkpoint size comparison

The official ModelScope base checkpoint is substantially larger:

| Checkpoint | Size |
|---|---:|
| `Fun-ASR-Nano-2512/model.pt` | 2,127,426,538 bytes |
| `Fun-ASR-MLT-Nano-2512/model.pt` | 1,971,162,898 bytes |

The difference is 156,263,640 bytes, which appears consistent with the missing
CTC decoder/head parameters.

## Expected behavior

One of the following:

1. Publish an MLT checkpoint containing trained `ctc_decoder.*` and `ctc.*`
   weights so the documented timestamp feature works for the 31-language model.
2. If timestamps are not supported by this checkpoint, fail closed instead of
   returning timestamps from randomly initialized layers, and document the
   limitation.

Could you confirm whether a trained multilingual CTC checkpoint exists or
whether the MLT checkpoint was packaged without those weights?

### PR DESCRIPTION
**Pull request by LauraGPT**
## Summary

- fail closed when a Fun-ASR-Nano checkpoint omits or only partially initializes `ctc_decoder.*` or `ctc.*`
- track the checkpoint keys that were actually loaded, including shape-mismatch handling
- apply the same guard to native, remote-code, and vLLM loading paths
- document the current `Fun-ASR-MLT-Nano-2512` timestamp limitation

## Root cause

The generic pretrained loader starts from the model's initialized state dict and replaces only compatible checkpoint tensors. Missing CTC tensors therefore remained randomly initialized while the final strict load still succeeded, so timestamp inference could run with untrained layers. The vLLM path had the same fail-open behavior.

This change disables all CTC timestamp components unless every required CTC tensor was loaded. Text transcription remains available.

Fixes #3208.

## Validation

- `PYTHONPATH="$PWD" python -m pytest -q tests/test_fun_asr_nano*.py`: **26 passed**
- `python -m py_compile` on all changed Python files
- vLLM decoder/head shape-mismatch regression coverage
- DDP-prefixed and wrapped checkpoint regression coverage
- Black 24.4.0 checks on the new/fully formatted files and `git diff --cached --check`
- Real `Fun-ASR-MLT-Nano-2512` checkpoint: 1,261 tensors, no CTC tensors; transcription succeeds and timestamp fields are absent
- Real ModelScope `Fun-ASR-Nano-2512` checkpoint: 1,347 tensors with 84 decoder and 2 CTC tensors; CTC remains enabled and timestamps are returned

The vLLM checkpoint-loading behavior is covered by focused unit tests; no CUDA vLLM end-to-end run was performed in this CPU validation environment.

### CODE REVIEW

File: funasr/train_utils/load_pretrained_model.py [hunk 1]
```diff
@@ -96,8 +97,12 @@ def load_pretrained_model(
...
             else:
                 dst_state[k] = src_state[k_src]
+                loaded_keys.add(k)
         else:
             print(f"Warning, miss key in ckpt: {k}, {path}")
 
     flag = obj.load_state_dict(dst_state, strict=True)
+    on_loaded = getattr(obj, "on_pretrained_model_loaded", None)
+    if callable(on_loaded):
+        on_loaded(frozenset(loaded_keys))
```
[the remark this answers is not in the document]
Comment (LauraGPT) [reply funasr/train_utils/load_pretrained_model.py:113]: Handled in e3ed3e97. The loader now falls back to `obj.module.on_pretrained_model_loaded` and removes one `module.` prefix only when calling the inner hook. `test_pretrained_loader_calls_inner_hook_with_clean_ddp_keys` covers this path.

File: funasr/models/fun_asr_nano/inference_vllm.py [hunk 2]
```diff
@@ -365,6 +354,31 @@ def _load_audio_components(self, model_dir: str, **kwargs):
...
+            key[len("ctc.") :]: value
+            for key, value in state_dict.items()
+            if key.startswith("ctc.") and not key.startswith("ctc_decoder.")
+        }
+        if ctc_state:
+            self.ctc.load_state_dict(ctc_state, strict=False)
+
+        loaded_keys = {f"ctc_decoder.{key}" for key in ctc_dec_state}
+        loaded_keys.update(f"ctc.{key}" for key in ctc_state)
+        disable_incomplete_ctc(self, loaded_keys, log=logger)
```
[the remark this answers is not in the document]
Comment (LauraGPT) [reply funasr/models/fun_asr_nano/inference_vllm.py:384]: Handled in e3ed3e97. vLLM now normalizes the checkpoint before key/shape filtering, including one DDP `module.` prefix and common wrapped states. The parameterized `ddp` / `wrapped` regression test covers both forms.

File: funasr/models/fun_asr_nano/model.py [hunk 3]
```diff
@@ -173,6 +175,11 @@ def __init__(
             if init_param_path is not None:
                 src_state = torch.load(init_param_path, map_location="cpu")
                 flag = self.ctc_decoder.load_state_dict(src_state, strict=False)
+                self._externally_loaded_ctc_keys.update(
+                    f"ctc_decoder.{key}"
+                    for key in self.ctc_decoder.state_dict()
+                    if key in src_state
+                )
```
[the remark this answers is not in the document]
Comment (LauraGPT) [reply funasr/models/fun_asr_nano/model.py:184]: Handled in e3ed3e97. Native Fun-ASR-Nano now normalizes the external decoder checkpoint before `load_state_dict`, recursively unwrapping `state_dict`, `model_state_dict`, or `model` and removing one DDP prefix. The shared normalizer has focused regression coverage.

File: examples/industrial_data_pretraining/fun_asr_nano/model.py [hunk 4]
```diff
@@ -137,6 +139,11 @@ def __init__(
             if init_param_path is not None:
                 src_state = torch.load(init_param_path, map_location="cpu")
                 flag = self.ctc_decoder.load_state_dict(src_state, strict=False)
+                self._externally_loaded_ctc_keys.update(
+                    f"ctc_decoder.{key}"
+                    for key in self.ctc_decoder.state_dict()
+                    if key in src_state
+                )
```
[the remark this answers is not in the document]
Comment (LauraGPT) [reply examples/industrial_data_pretraining/fun_asr_nano/model.py:151]: Handled in e3ed3e97. The remote-code model uses the same tested checkpoint normalizer as the native and vLLM paths before loading external CTC decoder weights.
